### PR TITLE
fix: modernize codebase with Go 1.26 patterns and address linter issues

### PR DESCRIPTION
**Pull request by tphakala**
## Summary

Modernizes the codebase to use Go 1.26 features and resolves all 30 gocritic/ruleguard warnings. Includes critical security fixes for path traversal vulnerabilities and IPv6 compatibility improvements.

### 🔒 High Priority: Security Fixes (10 files)

- **Path Traversal Prevention**: Replace manual `..` checks with `filepath.IsLocal()`
- **CVE Mitigations**: CVE-2023-45284 (space-padded reserved names), CVE-2023-45283 (\??\ prefix attacks)
- **Comprehensive Validation**: Detects Windows reserved names (COM1, LPT1, NUL, etc.)
- **Files**: api/v2, backup/targets, myaudio, security, support

**Why this matters**: Manual string checks miss edge cases that `filepath.IsLocal()` catches platform-specifically.

### 🌐 High Priority: IPv6 Compatibility (2 files)

- Replace `fmt.Sprintf("%s:%d", host, port)` with `net.JoinHostPort()`
- **Fixes**: Proper IPv6 address bracketing for FTP/SFTP connections
- **Before**: `2001:4860:4860::8888:443` (invalid)
- **After**: `[2001:4860:4860::8888]:443` (correct)
- **Files**: internal/backup/targets/{ftp,sftp}.go

### ⚡ Performance: errors.AsType() (9 locations)

- Migrate to `errors.AsType[T]()` for type-safe error assertions
- **Benefits**: ~3x faster, no reflection, compile-time type safety
- **Files**: api/v2 tests, internal/errors, internal/logger

**Before**:
```go
var httpErr *echo.HTTPError
if errors.As(err, &httpErr) {
    // use httpErr
}
```

**After**:
```go
if httpErr, ok := errors.AsType[*echo.HTTPError](err); ok {
    // use httpErr
}
```

### 🧪 Test Infrastructure (6 files)

- Replace `os.MkdirTemp()` with `t.TempDir()` for automatic cleanup
- Modernize to `math/rand/v2` (Go 1.22+) with `Int64N()` API
- **Benefits**: Auto-cleanup on test completion, better test isolation

### 📚 Documentation

- Enhanced `internal/CLAUDE.md` with Go 1.26 patterns and usage guidelines
- Added important caveats about `filepath.IsLocal()` path cleaning behavior
- Documented cross-platform impact analysis (Linux/Windows/macOS)
- Added golangci-lint rules for automated enforcement

## Verification

- ✅ **golangci-lint**: 0 issues (30/30 fixed)
- ✅ **Tests**: All passing with no regressions
- ✅ **Security**: Path traversal tests verified
- ✅ **Cross-platform**: Safe for Linux/Windows/macOS

## Testing Performed

```bash
# Linter check
golangci-lint run
# Output: 0 issues.

# Security tests
go test -v ./internal/security -run TestIsSafePath
# Output: PASS (26/26 tests)

# API v2 security scenarios
go test -v ./internal/api/v2 -run TestMediaSecurityScenarios
# Output: PASS (path traversal prevention verified)
```

## Breaking Changes

None. All changes are backward compatible.

## Related Issues

- Resolves gocritic/ruleguard warnings from golangci-lint 2.9.0 upgrade (#1956)
- Addresses code modernization for Go 1.26

## Summary by CodeRabbit

* **New Features**
  * Added extensive static-analysis rules and migration guidance for modern Go patterns (reflection iterators, rand v2, slice/bytes handling, crypto deprecations, host:port formatting) and test artifact handling.

* **Bug Fixes**
  * Strengthened path validation across components to better prevent traversal and platform edge cases; improved IPv6-safe address formatting.

* **Documentation**
  * Expanded Go best practices: testing (artifact dirs, parallelization, cleanup), preallocation, mocks, benchmarks, and I/O notes.

* **Tests**
  * Switched tests to framework-managed temp directories and updated test guidelines.

* **Chores**
  * Adopted type-safe error assertion recommendations and lint annotations.

### CODE REVIEW

File: rules/builtins.go [hunk 1]
```diff
@@ -165,3 +166,34 @@ func AppendWithoutValues(m dsl.Matcher) {
...
+//
+// See: https://go.dev/doc/go1.26#language
+func NewWithExpression(m dsl.Matcher) {
+	// Pattern: &[]T{v}[0] - the well-known slice hack for pointer-to-value
+	m.Match(
+		`&[]$typ{$val}[0]`,
+	).
+		Report("use new($val) instead of &[]$typ{$val}[0] (Go 1.26+)").
+		Suggest("new($val)")
+}
```
[the remark this answers is not in the document]
Comment (tphakala) [reply rules/builtins.go:201]: Good catch on the type safety issue. The Go 1.26 `new(expr)` feature uses the expression's type, so `new(42)` returns `*int`, not `*uint`. Changed to report-only without autofix since we can't reliably determine when type conversion is needed. Manual verification required.

For context: `new(expr)` where expr is a typed variable (e.g., `var x uint = 42`) correctly preserves type. The issue only occurs with untyped constants or explicit type conversions.

File: rules/crypto.go [hunk 2]
```diff
@@ -133,3 +133,45 @@ func DeprecatedRSAMultiPrime(m dsl.Matcher) {
...
+//
+// Recommended alternatives:
+//
+//	// OAEP encryption (preferred):
+//	ciphertext, _ := rsa.EncryptOAEP(sha256.New(), rand.Reader, pub, plaintext, nil)
+//	plaintext, _ := rsa.DecryptOAEP(sha256.New(), rand.Reader, priv, ciphertext, nil)
+//
+//	// OAEP with separate MGF1 hash (Go 1.26+):
+//	ciphertext, _ := rsa.EncryptOAEPWithOptions(rand.Reader, pub, plaintext, nil,
+//	    &rsa.OAEPOptions{Hash: crypto.SHA256, MGFHash: crypto.SHA1})
```
[the remark this answers is not in the document]
Comment (tphakala) [reply rules/crypto.go:154]: Fixed. The function signature is `EncryptOAEPWithOptions(random io.Reader, pub *PublicKey, msg []byte, opts *OAEPOptions)` - 4 parameters, not 5. Removed the extra `nil` argument.

File: rules/net.go [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+//   - Prevents directory traversal attacks
+//
+// See: https://pkg.go.dev/path/filepath#IsLocal
+func FilepathIsLocal(m dsl.Matcher) {
+	// Detect simple .. check that might be replaced by IsLocal
+	m.Match(
+		`strings.Contains($path, "..")`,
+	).
+		Report("consider using filepath.IsLocal($path) for comprehensive path validation (Go 1.20+)")
+}
```
[the remark this answers is not in the document]
Comment (tphakala) [reply rules/net.go:72]: Valid point - my own CLAUDE.md documentation recommends `strings.Contains(urlPath, "..")` for URL paths since `filepath.IsLocal` cleans paths internally.

Updated the rule message to clarify: "for file path validation (Go 1.20+); for URL paths, strings.Contains is appropriate". This informs developers that both patterns are valid depending on context.

The rule now educates rather than prescribes.

File: rules/slices.go [hunk 4]
```diff
@@ -63,88 +63,84 @@ func SortInts(m dsl.Matcher) {
...
-	// Pattern: append(b[:0:0], b...)
+	// append(s[:0:0], s...) pattern
 	m.Match(
-		`append($b[:0:0], $b...)`,
+		`append($s[:0:0], $s...)`,
 	).
-		Where(m["b"].Type.Is("[]byte")).
-		Report("use bytes.Clone($b) instead of append($b[:0:0], $b...) (Go 1.20+)")
+		Report("use slices.Clone($s) instead of append($s[:0:0], $s...) (Go 1.21+)")
 }
```
[the remark this answers is not in the document]
Comment (tphakala) [reply rules/slices.go:149]: Fixed. Added type guards to SlicesClone:
- `.Where(m["typ"].Text != "byte")` for the []T(nil) and []T{} patterns
- `.Where(!m["s"].Type.Is("[]byte"))` for the [:0:0] pattern

Now BytesClone and SlicesClone are mutually exclusive - byte slices only trigger BytesClone.

File: internal/backup/targets/common.go [hunk 5]
```diff
@@ -394,9 +394,9 @@ func ValidatePathWithOpts(pathToCheck string, opts PathValidationOpts) (string,
 		}
 	}
 
-	// Check for directory traversal
-	if strings.Contains(clean, "..") {
-		return "", backup.NewError(backup.ErrSecurity, "path contains directory traversal", nil)
+	// Use filepath.IsLocal for comprehensive path validation (prevents CVE-2023-45284, CVE-2023-45283)
+	if !filepath.IsLocal(clean) {
+		return "", backup.NewError(backup.ErrSecurity, "path is not local or contains traversal", nil)
 	}
```
[the remark this answers is not in the document]
Comment (tphakala) [reply internal/backup/targets/common.go:400]: Acknowledged as a minor UX concern. The behavior is safe - empty paths are correctly rejected. The error message could be clearer but doesn't affect security.

For context: This path is used by backup targets (FTP/SFTP/local filesystem). Empty paths after cleaning are invalid for backups regardless of the error message wording. The `filepath.IsLocal` check serves its primary purpose: preventing CVE-2023-45284/45283.

Low priority for refinement - current behavior is correct and secure.
